applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp

```diff
@@ -1039,6 +1164,46 @@ void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::SelectUpwindEle
     }
 }
 
+template<int TDim, int TNumNodes>
+int TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::GetAdditionalNode() const
+{
+    const TransonicPerturbationPotentialFlowElement& r_this = *this;
+
+    // current and upwind element geometry
+    const GeometryType& r_geom = r_this.GetGeometry();
+    const GeometryType& r_upwind_geom = mpUpwindElement->GetGeometry();
+
+    int non_matching_nodes_counter = 0;
+    int additional_node = -1;
+    bool loop_stop = false;
+
+    for (unsigned int i = 0; i < TNumNodes && !loop_stop; i++)
+    {
+        for (unsigned int j = 0; j < TNumNodes && !loop_stop; j++)
+        {
+            // check if nodes are the same
+            if(r_upwind_geom[i].Id() != r_geom[j].Id())
+            {
+                non_matching_nodes_counter++;
+            }
+        }
+        // get index of non matching node
+        if (non_matching_nodes_counter == TNumNodes)
+        {
+            additional_node = i;
+            loop_stop = true;
+        }
+        non_matching_nodes_counter = 0;
+    }
+
+    if(additional_node == -1)
+    {
+        KRATOS_WARNING("GetAdditionalNode") << " did not find additional node for element # " << r_this.Id() << std::endl;
+    }
+
+    return additional_node;
+}
+
```

Comment:
This would be my idea to make it shorter and reuse your code:

```cpp
template<int TDim, int TNumNodes>
int TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::GetAdditionalUpwindNodeIndex() const
{
    // current and upwind element geometry
    const GeometryType& r_geom = this->GetGeometry();
    const GeometryType& r_upwind_geom = pGetUpwindElement()->GetGeometry();
    std::vector<size_t> element_nodes_ids;
    PotentialFlowUtilities::GetSortedIds<TDim, TNumNodes>(element_nodes_ids, r_geom);

    // Search for the Id of the upwind element node that
    // is not contained in the current element
    bool upstream_element_id_found = false;
    // Loop over upwind element nodes
    for (unsigned int i = 0; i < TNumNodes; i++) {

        if (std::find(element_nodes_ids.begin(), element_nodes_ids.end(),
                      r_upwind_geom[i].Id()) == element_nodes_ids.end()) {
            upstream_element_id_found = true;
            return i;
        }
    }
    KRATOS_ERROR_IF(!upstream_element_id_found) << "No upstream element id found for element #" << this->Id() << std::endl;

    return -1;
}
```